starlette/middleware/base.py

```diff
@@ -4,14 +4,33 @@
 
 from starlette.requests import Request
 from starlette.responses import Response, StreamingResponse
-from starlette.types import ASGIApp, Receive, Scope, Send
+from starlette.types import ASGIApp, Message, Receive, Scope, Send
 
 RequestResponseEndpoint = typing.Callable[[Request], typing.Awaitable[Response]]
 DispatchFunction = typing.Callable[
     [Request, RequestResponseEndpoint], typing.Awaitable[Response]
 ]
 
 
+class _CachedRequest(Request):
+    async def wrapped_receive(self) -> Message:
+        if getattr(self, "_body", None) is not None:
+            # body() was called
+            return {
+                "type": "http.request",
+                "body": self._body,
+                "more_body": False,
+            }
+        elif self._stream_consumed:
+            # stream() was called to completion
+            return {
+                "type": "http.request",
+                "body": b"",
+                "more_body": False,
+            }
+        return await self.receive()
```

Comment:
I think handling here should be more stateful. In particular:
-  if the stream gets consumed by the downstream app we should mark the stream as consumed on this app instance
- if we've already retuned all of the body we should move onto polling receive() again (so that we can wait for the client to disconnect, etc)